core/arch/arm/kernel/secure_partition.c

```diff
@@ -1,50 +1,255 @@
 // SPDX-License-Identifier: BSD-2-Clause
 /*
- * Copyright (c) 2020, Arm Limited.
+ * Copyright (c) 2020-2021, Arm Limited.
  */
 #include <crypto/crypto.h>
 #include <initcall.h>
-#include <kernel/secure_partition.h>
+#include <kernel/ldelf_loader.h>
 #include <kernel/embedded_ts.h>
+#include <kernel/secure_partition.h>
 #include <kernel/ts_store.h>
+#include <ldelf.h>
+#include <mm/core_mmu.h>
+#include <mm/fobj.h>
+#include <mm/mobj.h>
+#include <mm/vm.h>
+#include <optee_ffa.h>
 #include <stdio.h>
 #include <string.h>
+#include <tee_api_types.h>
 #include <trace.h>
+#include <types_ext.h>
 #include <utee_defines.h>
 #include <util.h>
 #include <zlib.h>
 
+static const struct ts_ops *_sp_ops;
+const uint16_t MIN_ENDPOINT_ID = 0x8001;
+
+/* List that holds all of the loaded SP's */
+static struct sp_sessions_head open_sp_sessions =
+	TAILQ_HEAD_INITIALIZER(open_sp_sessions);
+
+static struct ts_ops sp_ops __rodata_unpaged = {
+	.enter_open_session = NULL,
+	.enter_invoke_cmd = NULL,
+	.enter_close_session = NULL,
+	.dump_state = NULL,
+	.destroy = NULL,
+	.get_instance_id = NULL,
+	.handle_svc = NULL,
+};
+
 static const struct embedded_ts *find_secure_partition(const TEE_UUID *uuid)
 {
 	const struct embedded_ts *sp = NULL;
 
-	for_each_secure_partition(sp)
+	for_each_secure_partition(sp) {
 		if (!memcmp(&sp->uuid, uuid, sizeof(*uuid)))
 			return sp;
+	}
+	return NULL;
+}
+
+bool is_sp_ctx(struct ts_ctx *ctx)
+{
+	return ctx && (ctx->ops == _sp_ops);
+}
+
+static void set_sp_ctx_ops(struct ts_ctx *ctx)
+{
+	ctx->ops = _sp_ops;
+}
+
+struct sp_session *sp_get_session(uint32_t session_id)
+{
+	struct sp_session *s = NULL;
+
+	TAILQ_FOREACH(s, &open_sp_sessions, link) {
+		if (s->endpoint_id == session_id)
+			return s;
+	}
 
 	return NULL;
 }
 
-static TEE_Result secure_partition_open(const TEE_UUID *uuid,
-					struct ts_store_handle **h)
+/*
+ * When starting the SP for the first time a init_info struct is passed.
+ * Store the struct on the stack and store the address in x0
+ */
+static void sp_init_info(struct sp_ctx *ctx)
 {
-	return emb_ts_open(uuid, h, find_secure_partition);
+	struct sp_ffa_init_info *info = NULL;
+
+	/* Store the secure partition info on top of stack. */
+	ctx->uctx.stack_ptr -= ROUNDUP(sizeof(*info), STACK_ALIGNMENT);
+
+	ctx->sp_regs.x[0]  = (uintptr_t)ctx->uctx.stack_ptr;
+	info = (struct sp_ffa_init_info *)ctx->uctx.stack_ptr;
+
+	info->magic = 0;
+	info->count = 0;
 }
 
-REGISTER_SP_STORE(2) = {
-	.description = "SP store",
-	.open = secure_partition_open,
-	.get_size = emb_ts_get_size,
-	.get_tag = emb_ts_get_tag,
-	.read = emb_ts_read,
-	.close = emb_ts_close,
-};
+static uint32_t new_session_id(struct sp_sessions_head *open_sessions)
+{
+	struct sp_session *last = NULL;
+	uint32_t id = MIN_ENDPOINT_ID + 1;
```

Comment:
My Apologies in that there is no major issue here.
I thought the overflow FFFF+1 == 0 would not be caught as out-of-memory new session ID.

However, I think the way the ID overflow gets detected is not obvious for one reading the code. Using `uint16_t` in this function would clarify. Unless other reviewers find it ok as-is.

Reply:
I still believe `id` here and return value for `new_session_id()` are `uint16_t` so that clearly returns 0 (failure as per caller) when increment below overflows.